File: Code_folder/GeneralRevision/decodestring.js
function isDigit(str) {

    // This checks if a single character is a digit (0-9).
    // ^\d$ ensures it matches exactly one digit, not multiple
    return /^\d$/.test(str);
}

function decodeString(stringA) {
    let stack = [];

    for (let i = 0; i < stringA.length; i++) {
        if (stringA[i] !== ']') {
            stack.push(stringA[i]);
        } else {
            // Extract the substring inside the brackets
            let subStr = "";
            while (stack.length > 0 && stack[stack.length - 1] !== '[') {
                subStr = stack.pop() + subStr;
            }

            // Pop the '[' from the stack
            stack.pop();

            // Extract the number (k)
            let k = "";
            while (stack.length > 0 && isDigit(stack[stack.length - 1])) {
                k = stack.pop() + k;
            }

            // Repeat the substring and push back to stack
            stack.push(subStr.repeat(parseInt(k)));
        }
    }

    return stack.join('');
}

const input = "3[a]2[bc]";
console.log(decodeString(input));  // Output: "aaabcbc"

File: Code_folder/GeneralRevision/Decode_String.py
def decodeString(s):
    stack = []
    # for i in range(len(s)):
    #     char = s[i]
    for char in s:
    # for i in range(len(s)):
    #     char = s[i]
        if char != ']':
            stack.append(char)
        else:
            # Get the string inside brackets
            substr = ''
            while stack and stack[-1] != '[':
                substr = stack.pop() + substr
            stack.pop()  # remove the '['

            # Get the repeat count (can be more than one digit)
            k = ''
            while stack and stack[-1].isdigit():
                k = stack.pop() + k

            # Repeat and push back
            stack.append(substr * int(k))

    return ''.join(stack)

# Test
input_str = "3[a]2[bc]"
print(decodeString(input_str))  # Output: aaabcbc
